### PR TITLE
dojo quick changes

### PR DESCRIPTION
**Pull request by Khady**
This PR is based on the work done during the recent dojo. It should probably have been divided into different pull requests, as commits are mostly independent from each other. Most of the changes are relatively small. 2 of them could be controversial though:

- support for external references (URI with or without a http prefix)
- generation of one atd file from a batch of schemas

Those two features kind of go hand in hand and are going in the same direction. 

I was trying jsonschema2atd on schemas that have http uri references to other types. In the past I had to update all of those by hands to turn them into `#/$defs/typename`. Now those references don't trigger an error. But it means that we can generate a atd file that wouldn't compile without manual fixes.

The generation from a batch of schemas helps with that, as I now can work with 3 schemas that related to each other. Schema A has http references to schemas B and C. So I can download those 3 files and then run `jsonschema2atd --format=jsonschema c.json b.json a.json`. There are cases which won't work, for example if 2 of the schemas have types named the same. But I'm not sure how to fix that without creating too much noise.

### CODE REVIEW

File: lib/generator.ml [hunk 1]
```diff
@@ -222,9 +238,10 @@ let make_atd_of_jsonschema input =
   let schema = Json_schema_j.schema_of_string input in
   let root_type_name = Option.value ~default:"root" schema.title in
   let defs =
-    match schema.defs with
-    | None -> []
-    | Some defs -> List.map (fun (name, schema) -> name, Obj schema) defs
+    let defs = List.concat [ Utils.list_of_nonempty schema.defs; Utils.list_of_nonempty schema.definitions ] in
+    match defs with
+    | [] -> []
+    | defs -> List.map (fun (name, schema) -> name, Obj schema) defs
```
Comment (ixzzd) [inline lib/generator.ml:244]: I guess we can do it like this
```suggestion
    List.map (fun (name, schema) -> name, Obj schema) defs
```

Comment (Khady) [reply lib/generator.ml:244]: :facepalm: yes you are right, thanks!

**Review by ixzzd — APPROVED** [review]
Looks great, thank you 👍

File: lib/generator.ml [hunk 2]
```diff
@@ -187,23 +226,42 @@ and process_one_of ~ancestors (schemas_or_refs : schema or_ref list) =
...
 
-and process_enums enums =
+and process_string_enums enums =
+  let enums =
+    List.map
+      (function
+        | `String s -> s
+        | value ->
+          failwith
+            (sprintf "Invalid value %s in string enum %s" (Yojson.Basic.to_string value)
```
Comment (ixzzd) [inline lib/generator.ml:236]: 👍

File: bin/main.ml [hunk 3]
```diff
@@ -12,18 +12,29 @@ module Input_format = struct
...
+    match paths with
+    | [ _ ] -> `Default
+    | _ -> `Per_file
+  in
+  List.iter
+    (fun path ->
+      let root =
+        match root with
+        | `Default -> None
+        | `Per_file -> Some (path |> Filename.basename |> Filename.remove_extension |> Utils.sanitize_name)
```
Comment (ixzzd) [inline bin/main.ml:32]: minor: wondering if it does make sense to use filename for root time always (even if only one input scheme file passed)
I used `jsonscheme2atd` for openapi files only, so I didn't have issues with root type name. @Khady if you think it is handy, feel free to change the default `root` type to filename

Comment (ixzzd) [reply bin/main.ml:32]: Also it would simplify code (we can get rid of `Default` and `Per_file` variants)

**Comment by Khady** [comment]
> I guess the relatively easy way would be to use global input_toplevel_schemas to resolve typename collisions (convert it to map probably? with input filenames as a key). Please open an issue if you face the this problem in real world usage

the problem with this solution is that one needs to create some kind of connection between the http uri references in the schemas and the actual json files that are passed to the cli. I was scared that doing it explicitly through the cli would make the invocations too noisy. We could build something automatic based on the `id`/`$id` of a schema probably.

File: lib/generator.ml [hunk 4]
```diff
@@ -15,23 +23,34 @@ let process_int_type schema =
...
+  in
+  let name_of_path path =
+    match path |> String.split_on_char '/' |> List.rev |> List.hd with
+    | exception _ -> failwith (sprintf "Unsupported ref value: %s" ref)
+    | name -> name
+  in
+  match pointer with
+  | None -> name_of_path uri
+  | Some pointer ->
+  match String.split_on_char '/' pointer with
```
Comment (Khady) [inline lib/generator.ml:40]: I'm guessing this is not completely correct. If the pointer is to a subschema then the name shouldn't be just the last part of the pointer

**Pull request merged**